pymeasure/adapters/protocol.py

```diff
@@ -30,15 +30,76 @@
 log.addHandler(logging.NullHandler())
 
 
+def to_bytes(command):
+    """Change `command` to a bytes object"""
+    if isinstance(command, (bytes, bytearray)):
+        return command
+    elif command is None:
+        return b""
+    elif isinstance(command, str):
+        return command.encode("utf-8")
+    elif isinstance(command, (list, tuple)):
+        return bytes(command)
+    elif isinstance(command, (int, float)):
+        return str(command).encode("utf-8")
+    raise TypeError(f"Invalid input of type {type(command).__name__}.")
+
+
 class ProtocolAdapter(Adapter):
     """ Adapter class testing command exchange without instrument hardware.
 
+    :param list of lists comm_pairs: List of message pairs. First message is
+        the write one, the second one is the read message.
+        'None' indicates that a pair member (write or read) does not exist.
     :param kwargs: TBD key-word arguments
     """
 
-    def __init__(self, preprocess_reply=None, **kwargs):
-        super().__init__(preprocess_reply=preprocess_reply)
-        # TODO: Make this skeleton implementation workable
+    def __init__(self, comm_pairs=[], preprocess_reply=None, **kwargs):
+        """Generate the adapter and initialize internal buffers."""
+        assert isinstance(comm_pairs, (list, tuple)), (
+            "Parameter comm_pairs has to be a list or tuple.")
+        super().__init__(preprocess_reply=preprocess_reply, **kwargs)
+        self._read_buffer = b""
+        self._write_buffer = b""
+        self.comm_pairs = comm_pairs
+        self._index = 0
+
+    def write(self, command):
+        """Compare the command with the expected one and fill the read."""
+        self.write_bytes(to_bytes(command))
+        assert self._write_buffer == b"", (
+            f"Written bytes '{self._write_buffer}' do not match expected "
+            f"'{self.comm_pairs[self._index][0]}'.")
+
+    def write_bytes(self, content):
+        """Write the bytes `content`. If a command is full, fill the read."""
+        self._write_buffer += content
+        p_write, p_read = self.comm_pairs[self._index]
+        if self._write_buffer == to_bytes(p_write):
+            assert self._read_buffer == b"", (
+                f"Unread response '{self._read_buffer}' present when writing. "
+                "Read the response or enable the property's 'check_set_errors'")
+            # Clear the write buffer
+            self._write_buffer = b""
+            self._read_buffer = to_bytes(p_read)
+            self._index += 1
+
+    def read(self):
+        """Return an already present or freshly fetched read buffer as a string."""
+        return (self.read_bytes(-1) + self.read_bytes(1)).decode("utf-8")
```

Comment:
I'll have to take some time to step through _why_ this works  as expected :D

Reply:
We're reading lists in this Adapter. - 1 returns all bytes but the last one, 1 returns the last byte.

A first implementation reading len(buffer) bytes failed, if the pair was (None, resp), because the length is zero in this case.